lib/e3db/client.rb

```diff
@@ -559,24 +621,226 @@ def revoke(type, reader_id)
+      end
+      ak = @ak_cache[cache_key][:ak]
+
+      encrypted_record = Record.new(meta: plain_record.meta, data: Hash.new)
+      plain_record.data.each do |k, v|
+        dk =   new_data_key
+        efN =  secret_box_random_nonce
+        ef =   RbNaCl::SecretBox.new(dk).encrypt(efN, v)
+
+        edkN = secret_box_random_nonce
+        edk =  RbNaCl::SecretBox.new(ak).encrypt(edkN, dk)
+
+        encrypted_record.data[k] = [edk, edkN, ef, efN].map { |f| base64encode(f) }.join(".")
+      end
+
+      encrypted_record
+    end
+
+    # Encrypt a new record consisting of the given data.
+    #
+    # +type+ is a string giving the record type. +data+ should be a
+    # dictionary of string values. +plain+ should be a dictionary of
+    # string values. +id+ should be the ID of the client creating the
+    # record. +eak+ should be an encrypted secret key, obtained via
+    # {#create_writer_key}.
+    #
+    # @return [Record] An instance containing the encrypted data.
+    def encrypt_new_record(type, data, plain, id, eak)
+      meta = Meta.new(record_id: nil, writer_id: id, user_id: id,
+                      type: type, plain: plain, created: nil,
+                      last_modified: nil, version: nil)
+
+      encrypt_record(Record.new(:meta => meta, :data => data), eak)
+    end
+
+    # Decrypts a record using the given secret key.
+    #
+    # The record should be either a JSON document (as a string) or a
+    # Record instance. +eak+ should be an encrypted secret key,
+    # obtained via {#get_reader_key}.
+    #
+    # @return [Record] An instance containing the decrypted data.
+    def decrypt_record(encrypted_record, eak)
+      encrypted_record = Record.new(JSON.parse(encrypted_record, symbolize_names: true)) if encrypted_record.is_a? String
+      raise 'Can only decrypt JSON string or Record instance.' if ! encrypted_record.is_a? Record
+
+      cache_key = [encrypted_record.meta.writer_id, encrypted_record.meta.user_id, encrypted_record.meta.type]
+      if ! @ak_cache.has_key? cache_key
+        @ak_cache[cache_key] = { :eak => eak, :ak => decrypt_box(eak.eak, eak.authorizer_public_key.curve25519, @private_key) }
+      end
+      ak = @ak_cache[cache_key][:ak]
+
+      plain_record = Record.new(data: Hash.new, meta: Meta.new(encrypted_record.meta))
+      encrypted_record[:data].each do |k, v|
+        edk, edkN, ef, efN = v.split('.', 4).map { |f| base64decode(f) }
+
+        dk = RbNaCl::SecretBox.new(ak).decrypt(edkN, edk)
+        pv = RbNaCl::SecretBox.new(dk).decrypt(efN, ef)
+
+        plain_record.data[k] = pv
+      end
+
+      plain_record
+    end
+
     private
 
+    # Returns the encrypted access key for this client for the
+    # given writer/user/type combination. Throws
+    # Faraday::ResourceNotFound if the key does not exist.
+    #
+    # Returns an instance of E3DB::EAK.
+    def get_eak(writer_id, user_id, type)
+      get_cached_key(writer_id, user_id, type)[:eak]
```

Comment:
As mentioned earlier, since `eak`s are specific to _readers_, will this interfere if multiple reader `eak`s are being cached? (Thinking about sharing ...)